R/data.table.R

```diff
@@ -2415,7 +2415,8 @@ split.data.table = function(x, f, drop = FALSE, by, sorted = FALSE, keep.by = TR
     if (!missing(by))
       stopf("passing 'f' argument together with 'by' is not allowed, use 'by' when split by column in data.table and 'f' when split by external factor")
     # same as split.data.frame - handling all exceptions, factor orders etc, in a single stream of processing was a nightmare in factor and drop consistency
-    return(lapply(split(x = seq_len(nrow(x)), f = f, drop = drop, ...), function(ind) x[ind]))
+    # be sure to use x[ind, ], not x[ind], in case downstream methods don't follow the same subsetting semantics (#5365)
+    return(lapply(split(x = seq_len(nrow(x)), f = f, drop = drop, ...), function(ind) x[ind, ]))
```

Comment:
Not sure if single column data.frame would be reduced to vector here if we don't add drop=FALSE? Cannot check atm

Reply:
Hmm, I also can't check ATM... it can't hurt to add `drop=FALSE` though, right?

Reply:
Ah, quite confusing that `split()` takes a `drop=` argument but it is "drop for factor levels" and not "drop for array dimensions" 😅 I was wondering why we'd supply `drop = drop` to the outer `split` call here instead of the inner `[`:

https://github.com/Rdatatable/data.table/blob/20b7f6a7716d67fce9c00b2253a310be46650386/R/data.table.R#L2472

Anyway, re: whether to suppl `drop = FALSE`, well, `split.data.frame` does that, and `!missing(f)` case is supposed to be the same as `split.data.frame()`:

https://github.com/r-devel/r-svn/blob/02984c3070ab53594ae80ebaa1b574e5daa2d3e8/src/library/base/R/split.R#L42

I couldn't come up with an example to trigger the issue anyway.